### PR TITLE
Remove async_trait from Handler

### PR DESCRIPTION
**Pull request by arpad-m**
Removes usage of async_trait from the `Handler` trait.

Related earlier work: #6305, #6464, #7303, #7342

### CODE REVIEW

File: pageserver/compaction/src/interface.rs [hunk 1]
```diff
@@ -141,18 +140,16 @@ pub trait CompactionLayer<K: CompactionKey + ?Sized> {
...
     where
         Self: 'a;
 
     /// Return all keys in this delta layer.
-    async fn load_keys<'a>(
+    fn load_keys<'a>(
         &self,
         ctx: &E::RequestContext,
-    ) -> anyhow::Result<Vec<Self::DeltaEntry<'_>>>;
+    ) -> impl Future<Output = anyhow::Result<Vec<Self::DeltaEntry<'_>>>> + Send;
```
Comment (conradludgate) [inline pageserver/compaction/src/interface.rs:152]: This has different argument capture semantics, but I guess it's fine. Edition 2024 can't come soon enough

**Review by conradludgate — APPROVED** [review]

File: pageserver/compaction/src/interface.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/compaction/src/interface.rs:152]: yeah there is small semantic changes but I think they are okay to live with and mostly a benefit (no boxing). None of the APIs are stable.

**Review by arssher — APPROVED** [review]

**Review by koivunej — APPROVED** [review]

**Comment by arpad-m** [comment]
There seems to be a pattern, it's always `test_wal_restore_initdb` failing at the end. However, it's already marked as a flaky test.

[debug-pg14](https://github.com/neondatabase/neon/actions/runs/8599276087/job/23562314382?pr=7301):

```
2024-04-08T12:32:58.9547417Z [gw0] [ 99%] FAILED test_runner/regress/test_wal_restore.py::test_wal_restore[debug-pg14] 
2024-04-08T13:13:14.6676068Z [gw13] [ 99%] FAILED test_runner/regress/test_wal_restore.py::test_wal_restore_initdb[debug-pg14] 
2024-04-08T13:13:14.6698227Z ##[error]The action has timed out.
```

[debug-pg15](https://github.com/neondatabase/neon/actions/runs/8599276087/job/23562314656?pr=7301):

```
2024-04-08T12:33:05.0833470Z [gw9] [ 99%] FAILED test_runner/regress/test_wal_restore.py::test_wal_restore[debug-pg15] 
2024-04-08T13:13:14.7419316Z [gw5] [ 99%] FAILED test_runner/regress/test_wal_restore.py::test_wal_restore_initdb[debug-pg15] 
2024-04-08T13:13:14.7442846Z ##[error]The action has timed out.
```

[debug-pg16](https://github.com/neondatabase/neon/actions/runs/8599276087/job/23562314930?pr=7301):

```
2024-04-08T12:33:43.3293132Z [gw13] [ 99%] FAILED test_runner/regress/test_wal_restore.py::test_wal_restore[debug-pg16] 
2024-04-08T13:13:14.3486077Z [gw0] [ 99%] FAILED test_runner/regress/test_wal_restore.py::test_wal_restore_initdb[debug-pg16] 
2024-04-08T13:13:14.3506386Z ##[error]The action has timed out.
```

**Comment by arpad-m** [comment]
Ok seems that `test_wal_restore_initdb` is failing quite reliably...

**Comment by arpad-m** [comment]
I've done some further [experiments](https://github.com/neondatabase/neon/compare/arpad/less_async_trait_investigation): it works until the commit with the move from `async_trait` to the native async in trait.

it might be the generics being handled differently now?

**Comment by arpad-m** [comment]
closing this as I want to focus on different things. In the future, at some point, it might make sense to revisit this but not for now.

**Pull request closed without merging**